lib/srv/alpnproxy/dialer.go

```diff
@@ -64,10 +65,7 @@ func (d ALPNDialer) DialContext(ctx context.Context, network, addr string) (net.
 		return nil, trace.BadParameter("missing TLS config")
 	}
 
-	var dialer ContextDialer = &net.Dialer{
-		KeepAlive: d.cfg.KeepAlivePeriod,
-		Timeout:   d.cfg.DialTimeout,
-	}
+	var dialer ContextDialer = apiclient.NewDialer(ctx, d.cfg.KeepAlivePeriod, d.cfg.DialTimeout)
 	if d.cfg.ALPNConnUpgradeRequired {
 		dialer = newALPNConnUpgradeDialer(d.cfg.KeepAlivePeriod, d.cfg.DialTimeout, d.cfg.TLSConfig.InsecureSkipVerify)
```

Comment:
The `d.cfg.ALPNConnUpgradeRequired` flow will overwrite the `apiclient.NewDialer` 
I think that we cloud pass the `dialer` to `newALPNConnUpgradeDialer` and assign to https://github.com/gravitational/teleport/blob/0ee91f6c37eb3a8c5bc98b15f857c95902df50b2/lib/srv/alpnproxy/conn_upgrade.go#L88 to support custom HTTPS proxy in `ALPNConnUpgradeRequired` flow. 

@greedy52 Could you also take a look ?

Reply:
Yes, i believe the connection upgrade flow needs fix.

If we are going to support other protocols for connection upgrades, ideally we should move the connection upgrade logic inside the api dialer, something like:
```
apiclient.NewDialer(&apiclient.DialerConfig{
  KeepAlivePeriod: ...,
  DialTiemout: ...,
  ALPNUpgradeRequired: ...,
  Insecure: ...,
})
```
Seems too big of a change for now.

So agreed with what @smallinsky suggested to replace the `net.Dialer` with api dialer inside `alpnConnUpgradeDialer` for now.

Reply:
Hey @greedy52 and @smallinsky, I've updated the code to pass in the `apiclient.ContextDialer`, and amended and passed the effected tests. Feel free to take a look when you get a chance 🙏 

A quick question, in [dailer.go:29](https://github.com/gravitational/teleport/blob/40a4d8946a7aec1be676233b24fbe87a6fd3a289/lib/srv/alpnproxy/dialer.go#L29) we are re-defining another ContextDialer interface, and now that `apiclient` is imported we have two interfaces of the same name (although different namespace/package name) with identical signatures, is this something that we think we can refactor?

Reply:
I'm also AFK for the next week or so so I won't be able to commit any changes til I'm back, but please feel free to leave some more review comments that I can address once I'm back! 

Thanks Team!

Reply:
> A quick question, in [dailer.go:29](https://github.com/gravitational/teleport/blob/40a4d8946a7aec1be676233b24fbe87a6fd3a289/lib/srv/alpnproxy/dialer.go#L29) we are re-defining another ContextDialer interface, and now that apiclient is imported we have two interfaces of the same name (although different namespace/package name) with identical signatures, is this something that we think we can refactor?

Yes, I think we can use the same one from apiclient (or maybe redefine like `type ContextDialer apiclient.ContextDialer`)